engine/modules/opentelemetry/src/opentelemetry.cc

```diff
@@ -0,0 +1,279 @@
+
+#include "otl_fmt.hh"
+#include "otl_server.hh"
+
+using namespace com::centreon::engine::modules::opentelemetry;
+
+/**
+ * @brief Construct a new open telemetry::open telemetry object
+ *
+ * @param config_file_path
+ * @param io_context
+ */
+open_telemetry::open_telemetry(
+    const std::string_view config_file_path,
+    const std::shared_ptr<asio::io_context>& io_context,
+    const std::shared_ptr<spdlog::logger>& logger)
+    : _config_file_path(config_file_path),
+      _logger(logger),
+      _second_timer(*io_context),
+      _io_context(io_context) {
+  SPDLOG_LOGGER_INFO(_logger, "load of open telemetry module");
+}
+
+/**
+ * @brief to call when configuration changes
+ *
+ */
+void open_telemetry::_reload() {
+  std::unique_ptr<otl_config> new_conf =
+      std::make_unique<otl_config>(_config_file_path, *_io_context);
+  if (!_conf || !(*new_conf->get_grpc_config() == *_conf->get_grpc_config())) {
+    this->_create_otl_server(new_conf->get_grpc_config());
+  }
+
+  if (!_conf || !(*_conf == *new_conf)) {
+    fmt::formatter<::opentelemetry::proto::collector::metrics::v1::
+                       ExportMetricsServiceRequest>::max_length_log =
+        new_conf->get_max_length_grpc_log();
+    fmt::formatter<::opentelemetry::proto::collector::metrics::v1::
+                       ExportMetricsServiceRequest>::json_grpc_format =
+        new_conf->get_json_grpc_log();
+
+    _conf = std::move(new_conf);
+  }
+}
+
+/**
+ * @brief static creator of singleton
+ *
+ * @param config_path
+ * @param io_context
+ * @return std::shared_ptr<open_telemetry>
+ */
+std::shared_ptr<open_telemetry> open_telemetry::load(
+    const std::string_view& config_path,
+    const std::shared_ptr<asio::io_context>& io_context,
+    const std::shared_ptr<spdlog::logger>& logger) {
+  if (!_instance) {
+    _instance =
+        std::make_shared<open_telemetry>(config_path, io_context, logger);
+    instance()->_reload();
+    instance()->_start_second_timer();
+  }
+  return instance();
+}
+
+/**
+ * @brief create grpc server witch accept otel collector connections
+ *
+ * @param server_conf json server config
+ */
+void open_telemetry::_create_otl_server(
+    const grpc_config::pointer& server_conf) {
+  try {
+    std::shared_ptr<otl_server> to_shutdown = std::move(_otl_server);
+    if (to_shutdown) {
+      to_shutdown->shutdown(std::chrono::seconds(10));
+    }
+    _otl_server = otl_server::load(
+        server_conf,
+        [me = shared_from_this()](const metric_request_ptr& request) {
+          me->_on_metric(request);
+        },
+        _logger);
+  } catch (const std::exception& e) {
+    SPDLOG_LOGGER_ERROR(_logger, "fail to create opentelemetry grpc server: {}",
+                        e.what());
+    throw;
+  }
+}
+
+/**
+ * @brief static method used to make singleton reload is configuration
+ *
+ */
+void open_telemetry::reload(const std::shared_ptr<spdlog::logger>& logger) {
```

Comment:
why the logger as parameter since open_telemetry embed its logger?

Reply:
static method